### DISCUSSION TITLE
Where does GitRepository clone the files

### DISCUSSION THREAD

**Discussion by peldenwong** [fluxcd/flux2#2965] [Q&A] [answered]
I am trying to extend the GitRepository controller to run a command, e.g. `touch example.yaml`, everytime it reconciles due to how templating works in my clusters. What this command does is it produces a new yaml file that determines the values based on the cluster's specs.

I am able to run this command, however, my question is, how do I edit and save this file in GitRepository's local copy.

**Comment by kingdonb** [comment, marked as answer]
My understanding of your inquiry is that you are trying to inject a file into the GitRepository artifact. You probably will not be able to do that, the directory is randomly generated and not at a predictable location; however if you explain more about your use case we can possibly help understand and suggest a resolution.

What kind of information goes in this YAML file exactly? And why is it that you're avoiding putting the file into git directly (I guess it's some detail that differs per-cluster, but is there anything that we don't know about your situation that explains why you can't have that file in a gitrepo somewhere?)

There are a number of strategies including kustomize bases and gitrepository includes that may help here, but without knowing more about your use case I'm not sure how to suggest the best approach. Can you provide more info?

**Reply by peldenwong** [reply]
In my git repository, there is a file and a separate program that gets used to create a `values.yaml` file, which defines the values to go into the templates. The created file is, as you have said, cluster dependent. This `values.yaml` file is needed when deploying the helm chart, and also during reconciliation. I thought the best way to make it work is to run a bash command to generate the file (how it's done right now), and inject it into the GitRepository artifact so that when the helm release reconciles, the `values.yaml` file is available to be used. 

Alternatively, I was thinking if it is better to generate this `values.yaml` file during helm release reconciliation.

**Reply by kingdonb** [reply]
Flux works to guarantee intact delivery of artifacts with no injections, what you're describing sounds closest to this doc which you may find handy if you have not seen it yet:

https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

You can keep `values.yaml` in a directory that's dedicated for that cluster, in the bootstrap directory for example alongside of `flux-system`. You can use configmapgenerator to put it in a HelmRelease and this is the best known way to use Flux with Helm, for a number of reasons explained there. Mainly, this is the only pattern that guarantees a rollout will accompany every config change. When the configmap is rewritten alongside of the HelmRelease, the Kustomize generator adjusts the name to include the hash suffix in every place where it occurs as a reference to the configmap in the HelmRelease, so every update gets accompanied by a new rollout. This works with Helm or Kustomize targets either way just as well.

This is how Flux is anticipated to be used, please try it out and see if it can fit your requirements. You may need to run the bash script ahead of time and capture the output for each cluster in a commit. You can do this in CI right now.

**Reply by peldenwong** [reply]
Thank you, this does sound more promising. However, the bash script that generates `values.yaml` also fetches credentials and tokens that will be regularly renewed, can this still be done with the above mentioned approach without having to manually run the bash script each time?

**Reply by kingdonb** [reply]
What are the credentials used for exactly? Flux works with external secrets operators too, but you may need another outside tool such as [Reloader](https://github.com/stakater/Reloader) to restart deployments when their secrets or configmaps get updated, since you will need the capacity to make sure those containers get restarted if they are to use environment variables for their secrets.

If you have such a token that needs to be regularly refreshed, one way you can handle this is with a job that runs in the cluster. I'm pretty sure we still have some examples of credential refresh token cronjobs in the docs under the Azure guide or somewhere. It depends on how they fetch the credentials and what KMS system is used, if there is one.

If it's a credential for ECR or ACR container registry for example, then Flux can already autologin those resources with an IAM role assignment through annotations, such that you do not have to write any job or add any additional processes.

It does break Flux's potential multi-tenancy isolation though somewhat
